juju/model.py

```diff
@@ -1912,6 +1918,111 @@ def charmstore(self):
         controller = await self.get_controller()
         return await controller.remove_offer(self.info.uuid, endpoint, force)
 
+    async def consume(self, endpoint, application_alias=""):
+        """
+        Adds a remote offer to the model. Relations can be created later using
+        "juju relate".
+        """
+        try:
+            offer = parse_offer_url(endpoint)
+        except OfferParseError as e:
+            log.error(e.message)
+            raise
+        if offer.has_endpoint():
+            raise JujuError("remote offer {} should not include an endpoint".format(endpoint))
+        if offer.user == "":
+            offer.user = self.info.username
+            endpoint = offer.string()
+
+        source = await self._get_source_api(offer)
+        consume_details = await source.get_consume_details(offer.as_local().string())
+        if consume_details is None or consume_details.offer is None:
+            raise JujuAPIError("missing consuming offer url for {}".format(offer.string()))
+
+        offer_url = parse_offer_url(consume_details.offer.offer_url)
+        offer_url.source = offer.source
+
+        consume_details.offer.offer_url = offer_url.string()
+        consume_details.offer.application_alias = application_alias
+
+        arg = _create_consume_args(consume_details.offer,
+                                   consume_details.macaroon,
+                                   consume_details.external_controller)
+
+        facade = client.ApplicationFacade.from_connection(self.connection())
+        return await facade.Consume(args=[arg])
+
+    async def remove_saas(self, name):
+        """
+        Removing a consumed (SAAS) application will terminate any relations that
+        application has, potentially leaving any related local applications
+        in a non-functional state.
+        """
+        if not is_valid_application(name):
+            raise JujuError("invalid SAAS application name {}".format(name))
+
+        arg = client.DestroyConsumedApplicationParams()
+        arg.application_tag = application_tag(name)
+
+        facade = client.ApplicationFacade.from_connection(self.connection())
+        return await facade.DestroyConsumedApplications(applications=[arg])
+
+    async def _get_source_api(self, url):
+        controller = Controller()
+        if url.source == "":
+            current = await self.get_controller()
+            url.source = current.controller_name
+        await controller.connect(controller_name=url.source)
+        return controller
+
+
+def _create_consume_args(offer, macaroon, controller_info):
+    """
+    Convert a typed object that has been normalised to a overrided typed
```

Comment:
can you provide an example thing here to get a visual idea of the transforming expected please?

Reply:
So the issue here is that macaroons aren't exposed correctly, so we have to fill in the blanks